File: Java/프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder("");
        int[][] keyPad = new int[][] 
        {{1, 2, 3},{4, 5, 6}, {7, 8, 9}, {-1, 0, -1}};
        int leftRow = 3;
        int leftColumn = 0;
        int rightRow = 3;
        int rightColumn = 2;
    
        for(int num : numbers) {
            if(num == 1 || num == 4 || num == 7) {
                answer.append("L");
                for(int i=0; i<keyPad.length; i++) {
                    if(keyPad[i][0] == num) {
                        leftRow = i;
                        leftColumn = 0;
                        break;
                    }
                }
            } else if(num == 3 || num == 6 || num == 9) {
                answer.append("R");
                for(int i=0; i<keyPad.length; i++) {
                    if(keyPad[i][2] == num) {
                        rightRow = i;
                        rightColumn = 2;
                        break;
                    }
                }
            } else {
                for(int i=0; i<keyPad.length; i++) {
                    if(keyPad[i][1] == num) {
                        int leftDis = Math.abs(i-leftRow) + Math.abs(1-leftColumn);
                        int rigthDis = Math.abs(i-rightRow) + Math.abs(1-rightColumn);
                        if(leftDis < rigthDis) {
                            answer.append("L");
                            leftRow = i;
                            leftColumn = 1;
                            break;
                        } else if(rigthDis < leftDis) {
                            answer.append("R");
                            rightRow = i;
                            rightColumn = 1;
                            break;
                        } else {
                            if(hand.equals("left")) {
                                answer.append("L");
                                leftRow = i;
                                leftColumn = 1;
                                break;
                            } else {
                                answer.append("R");
                                rightRow = i;
                                rightColumn = 1;
                                break;
                            }
                        }
                    }
                }
            }
        }
        return answer.toString();
    }
}

File: Kotlin/프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.kt
import kotlin.math.*
class Solution {
    var sq = StringBuilder()
    var leftRow = 3
    var leftCol = 0
    var rightRow = 3
    var rightCol = 2
    
    fun solution(numbers: IntArray, hand: String): String {
        var h = hand
        var keyPad = arrayOf(arrayOf(1,2,3),arrayOf(4,5,6),arrayOf(7,8,9),arrayOf(-1,0,-1))
        numbers.forEach{
            when {
                it==1 -> pickHand(0,0,"L")
                it==4 -> pickHand(1,0,"L")
                it==7 -> pickHand(2,0,"L")
                it==3 -> pickHand(0,2,"R")
                it==6 -> pickHand(1,2,"R")
                it==9 -> pickHand(2,2,"R")
                else -> {
                    for(i in 0..3) {
                        for(j in 0..2) {
                            if(keyPad[i][j]==it) {
                                var leftMin = abs(leftRow-i)+abs(leftCol-j)
                                var rightMin = abs(rightRow-i)+abs(rightCol-j)
                                if(leftMin<rightMin) {
                                    pickHand(i,j,"L")
                                } else if(rightMin<leftMin) {
                                    pickHand(i,j,"R")
                                } else {
                                    if(h=="right") {
                                        pickHand(i,j,"R")
                                } else {
                                        pickHand(i,j,"L")
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        return sq.toString()
    }
    fun pickHand(row :Int, col :Int, dir:String) {
        if(dir=="L") {
            leftRow = row
            leftCol = col
            sq.append("L")
        } else {
            rightRow = row
            rightCol = col
            sq.append("R")
        }
    }
}